packages/ddc-war/DDC/War/Interface/Controller.hs

```diff
@@ -81,10 +81,14 @@ controller config gang chainsTotal chanResult
                 gangState       <- getGangState gang
                 if gangState == GangFinished
 
-                 -- No job results in channel, and the gang has finished.
-                 -- We're done, so return to caller.
+                 -- Just because the gang loop has finished doesn't mean that all the workers have.
+                 -- Wait for the workers to finish, then read the remaining bits off the queue.
                  then do
-                        return jobResults
```

Comment:
race condition was here: the gang state was finished, but that doesn't mean the workers are finished, so the results of the last workers were usually being ignored